### PR TITLE
pattern-where-python

### LINKED ISSUE #101: subpattern where-python:`arbitrary python code`
We'd like to make rule files even more pluggable and dynamic. To that end, it would be valuable to have python / javascript scriptability that can execute arbitrary code to decide whether or not a rule should fire. For instance:

```
pattern-with-guard:
    pattern: foo($X)
    where-python: "0 <= vars['$X'] <= 100"
```

Specifically, many of the checks @mschwager wrote in dlint would be good candidates for this pattern. Eg https://github.com/dlint-py/dlint/blob/master/dlint/linters/bad_urllib3_kwarg_use.py could become:

```
pattern-with-guard:
    pattern: urllib3.PoolManager(cert_reqs=$X)
    where-python: "vars['$X'] in ['CERT_NONE', 'NONE']"
```

### PR DESCRIPTION
**Pull request by ievans**
cc @clintgibler

tests in https://github.com/returntocorp/sgrep-rules/pull/185 and https://github.com/returntocorp/sgrep-rules/pull/196

when complete, closes #101 ; see https://github.com/returntocorp/sgrep/pull/166 for prior discussion

Major changes include using proper types and classes for the sgrep output as well as `BooleanRuleExpression` for the formerly-tuple parsed expressions.

Because of the potential for arbitrary code execution, these rules are only run if the command-line flag `--dangerously-allow-arbitrary-code-execution-from-rules` is set

### CODE REVIEW

**Comment by ievans** [comment]
A confusing case:
```
      - pattern-not-inside: $CALLER(...)
      - pattern-where-python: "vars['$CALLER'] in ['assert', 'assertFalse', 'assertTrue']"
```

Will never match anything inside any function call, because "$CALLER(...)` captures and removes them before the filter can run.

**Comment by mschwager** [comment]
Would it be possible to split this into two separate PRs? 1.) for the refactoring, i.e. `BooleanRuleExpression`, `SgrepRange`, `SRange`, typing changes, etc, 2.) for building on (1) with `pattern-where-python` functionality. That would make review much easier and help ensure we don't miss anything.

File: sgrep.py [hunk 1]
```diff
@@ -229,77 +257,140 @@ def _evaluate_single_expression(
...
                     break
-        # print(f"after filter `{operator}`: {output_ranges}")
+        debug_print(f"after filter `{expression.operator}`: {output_ranges}")
+        return output_ranges
+    elif expression.operator == OPERATORS.WHERE_PYTHON:
+        if not RCE_RULE_FLAG not in flags:
+            print_error_exit(
+                f"at least one rule needs to execute arbitrary code; this is dangerous! if you want to continue, enable the flag: RCE_RULE_FLAG"
+            )
+        assert expression.operand, "must have operand for this operator type"
```
Comment (aryx) [inline sgrep.py]: alala, would be nice to have algebraic datatype in Python ...

File: sgrep.py [hunk 2]
```diff
@@ -229,77 +257,140 @@ def _evaluate_single_expression(
...
+) -> bool:
+    # TODO: filter out obvious dangerous things here
+    global output
+    output = None
+
+    # HACK: we're executing arbitrary Python in the where-python,
+    # be careful my friend
+    vars = metavars
+    try:
+        exec(f"global output; output = {where_expression}")
```
Comment (aryx) [inline sgrep.py]: woohoo

Comment (aryx) [reply sgrep.py]: What is the difference between exec and eval?

Comment (ievans) [reply sgrep.py]: https://stackoverflow.com/questions/2220699/whats-the-difference-between-eval-exec-and-compile/29456463#29456463 @aryx this is a great explanation

**Pull request merged**